mono/profiler/decode.c

```diff
@@ -3063,6 +3278,170 @@ dump_heap_shots (void)
+dump_coverage (void)
+{
+	if (!coverage_methods && !coverage_assemblies)
+		return;
+
+	gather_coverage_statements ();
+	fprintf (outfile, "\nCoverage Summary:\n");
+
+	if (coverage_outfile) {
+		fprintf (coverage_outfile, "<?xml version=\"1.0\"?>\n");
+		fprintf (coverage_outfile, "<coverage version=\"0.3\">\n");
+	}
+
+	g_ptr_array_sort (coverage_assemblies, sort_assemblies);
+
+	for (guint i = 0; i < coverage_assemblies->len; i++) {
+		CoverageAssembly *assembly = coverage_assemblies->pdata[i];
+		GPtrArray *classes;
+
+		if (assembly->number_of_methods != 0) {
+			int percentage = ((assembly->fully_covered + assembly->partially_covered) * 100) / assembly->number_of_methods;
+			fprintf (outfile, "\t%s (%s) %d%% covered (%d methods - %d covered)\n", assembly->name, assembly->filename, percentage, assembly->number_of_methods, assembly->fully_covered);
+		} else {
+			fprintf (outfile, "\t%s (%s) ?%% covered (%d methods - %d covered)\n", assembly->name, assembly->filename, assembly->number_of_methods, assembly->fully_covered);
+		}
+
+		if (coverage_outfile) {
+			char *escaped_name, *escaped_filename;
+			escaped_name = escape_string_for_xml (assembly->name);
+			escaped_filename = escape_string_for_xml (assembly->filename);
+
+			fprintf (coverage_outfile, "\t<assembly name=\"%s\" guid=\"%s\" filename=\"%s\" method-count=\"%d\" full=\"%d\" partial=\"%d\"/>\n", escaped_name, assembly->guid, escaped_filename, assembly->number_of_methods, assembly->fully_covered, assembly->partially_covered);
+
+			g_free (escaped_name);
+			g_free (escaped_filename);
+		}
+
+		classes = g_hash_table_lookup (coverage_assembly_classes, assembly->name);
+		if (classes) {
+			for (guint j = 0; j < classes->len; j++) {
+				CoverageClass *klass = classes->pdata[j];
+
+				if (klass->number_of_methods > 0) {
+					int percentage = ((klass->fully_covered + klass->partially_covered) * 100) / klass->number_of_methods;
+					fprintf (outfile, "\t\t%s %d%% covered (%d methods - %d covered)\n", klass->class_name, percentage, klass->number_of_methods, klass->fully_covered);
+				} else {
+					fprintf (outfile, "\t\t%s ?%% covered (%d methods - %d covered)\n", klass->class_name, klass->number_of_methods, klass->fully_covered);
+				}
+
+				if (coverage_outfile) {
+					char *escaped_name;
+					escaped_name = escape_string_for_xml (klass->class_name);
+
+					fprintf (coverage_outfile, "\t\t<class name=\"%s\" method-count=\"%d\" full=\"%d\" partial=\"%d\"/>\n", escaped_name, klass->number_of_methods, klass->fully_covered, klass->partially_covered);
+					g_free (escaped_name);
+				}
+			}
+		}
+	}
+
+	for (guint i = 0; i < coverage_methods->len; i++) {
+		CoverageMethod *method = coverage_methods->pdata[i];
+
+		/* Sanity check */
+		if (method->n_statements != method->coverage->len) {
+			fprintf (outfile, "Incorrect number of statements for %s:%s (%d expected, have %d)\n", method->class_name, method->method_name, method->n_statements, method->coverage->len);
```

Comment:
I'm not sure the assert should be in proflog.c, when the output.mlpd file is generated the two numbers being checked will always match, they're both generated from the same source. This check is more to check that the file was parsed correctly and didn't get corrupted somehow.

In reality, it might not be necessary, if the file got parsed incorrectly or corrupted, worse things will happen.